Speed up TypeScript projects

deftomat opened the pull request:
As a lot of [people](https://hackernoon.com/why-i-no-longer-use-typescript-with-react-and-why-you-shouldnt-either-e744d27452b4) is complaining about TypeScript performance in CRA, I decided to enable `async` mode in TypeScript checker.

These changes basically brings the JS compilation times to TS projects. So, recompilation took less than 1 second instead of 3 seconds in medium size project.

The problem with async mode is that type-errors are reported after Webpack ends up recompilation as TypeScript could be slower than Babel. PR allows to emit files compiled by Babel immediately and then wait for TS and show type errors in terminal later. Also, if there was no compilation errors and any type error occurs, we trigger a hot-reload with new errors to show error overlay in browser.

Also, I wanted to start a discussion about `skipLibCheck: false` option in default `tsconfig.json`. This makes recompilations really slow and we should consider to set it to `true` or at least give users a big warning to let them know that it could be really slow.

The following video is showing the updated workflow with a forced 2.5 second delay for type-check to give you an idea how it works.

![nov-26-2018 15-47-01](https://user-images.githubusercontent.com/5549148/49021284-9446fe80-f192-11e8-952b-8f83d77d5fbc.gif)


I'm pretty sure that PR needs some polishing and improvements but it should works as it is. Especially a "hack" with reloading the browser after type-check looks ugly to me.

cc @brunolemos as he is an initiator of an original TypeScript PR.

Should fix https://github.com/facebook/create-react-app/issues/5820

Timer commented:
That error actually occurs on purpose and is caught. :-)

deftomat commented:
@Timer I'm not sure if I understand. AppVeyor is failing as test is NOT expecting that it should fail on purpose.

Timer commented:
Hm, it should be running this code:
https://github.com/facebook/create-react-app/blob/81c0cd4b1906157d30408971b3849b29ecabb3b9/tasks/e2e-simple.sh#L218-L237

Ensuring the build fails but tests continue: `yarn build; test $? -eq 1 || exit 1`

Is this not working correctly?

deftomat commented:
Sorry, my bad. Build failed due to 60 minutes timeout.

Exit code was checked correctly:
![image](https://user-images.githubusercontent.com/5549148/49086790-a63da580-f255-11e8-8f1b-515942cfff79.png)

deftomat commented:
TypeScript formatter now supports messages from TSLint.

(Should make https://github.com/facebook/create-react-app/compare/master...ianschmitz:tslint easier)

deftomat commented:
`skipLibCheck` will be true by default. See https://github.com/facebook/create-react-app/pull/5959

deftomat commented:
Not sure how to fix the failing checks as AppVeyor is failing due to 60 minutes limit for this repo and Netlify details just redirects to the 404.

Anyway, I think that PR is ready for a review.

iansu commented:
Appveyor has been broken for a while so don't worry about that. And Netlify is just to preview docs changes.

brunolemos commented:
I'm taking a quick break from open source to finish my project so hopefully another person can review this

ianschmitz commented on packages/react-dev-utils/WebpackDevServerUtils.js:
```diff
@@ -152,9 +159,32 @@ function createCompiler(webpack, config, appName, urls, useYarn) {
+
+    if (useTypeScript && statsData.errors.length === 0) {
+      process.stdout.write(
+        chalk.yellow('Files successfully emitted, waiting for typecheck results...')
```
Is the thinking here that the dev server would already be in the process of refreshing the browser with the updated source at this point, but that the type checking could take a while longer?

deftomat replied:
Yes, it could flicker in a first build as TS is probably faster than Babel. But in the following rebuilds, Babel is almost always faster than TS. So, the idea is to emit files as quick as possible and remind the user that type-check is still in progress. Webpack actually emit files just before it call *done* hook.

ianschmitz commented on packages/react-dev-utils/WebpackDevServerUtils.js:
```diff
@@ -152,9 +159,32 @@ function createCompiler(webpack, config, appName, urls, useYarn) {
+    const statsData = stats.toJson({ all: false, warnings: true, errors: true });
+
+    if (useTypeScript && statsData.errors.length === 0) {
+      process.stdout.write(
```
If we keep this can we change to use `console.log` to be consistent with other logging in this file?

deftomat replied:
Well, the idea is to remove this line when TS finish the type-checking. I'm not sure how to do it with `console.log`. Any idea?

deftomat replied:
We can call `clearConsole` once again but with this approach, it is much easier to modify the code later as this `if` can basically clear after itself without loosing any other important text in terminal.

ianschmitz commented on packages/react-dev-utils/WebpackDevServerUtils.js:
```diff
@@ -194,6 +224,25 @@ function createCompiler(webpack, config, appName, urls, useYarn) {
+  if (isSmokeTest) {
+    compiler.hooks.failed.tap('smokeTest', async () => {
+      await tsMessagesPromise;
+      process.exit(1)
```
Reminder for myself to do a quick sanity on these

ianschmitz commented:
Hey @deftomat. Thanks for the PR! It's looking really good so far. I'm hoping I'll have time to give it a spin this weekend.

ianschmitz commented on packages/react-scripts/config/webpack.config.js:
```diff
@@ -640,7 +640,7 @@ module.exports = function(webpackEnv) {
           watch: paths.appSrc,
           silent: true,
-          formatter: typescriptFormatter,
+          formatter: isEnvProduction ? typescriptFormatter : undefined,
```
```suggestion
          // The formatter is invoked directly in WebpackDevServerUtils during development
          formatter: isEnvProduction ? typescriptFormatter : undefined,
```

deftomat commented on packages/react-scripts/scripts/start.js:
```diff
@@ -94,9 +94,14 @@ checkBrowsers(paths.appPath, isInteractive)
+    };
     // Create a webpack compiler that is configured with custom messages.
-    const compiler = createCompiler(webpack, config, appName, urls, useYarn);
+    const compiler = createCompiler(webpack, config, appName, urls, useYarn, useTypeScript, devSocket);
```
yes

deftomat commented on packages/react-dev-utils/webpackHotDevClient.js:
```diff
@@ -140,19 +140,15 @@ function handleWarnings(warnings) {
   // Attempt to apply hot updates or reload.
   if (isHotUpdate) {
     tryApplyUpdates(function onSuccessfulHotUpdate() {
-      // Only print warnings if we aren't refreshing the page.
```
Yes, we need to print then all the time as we can send them asynchronously after the hot reload.
Without this change, they can be ignored as there are no updates in code. Just a new warnings, so this callback will never run for them.

bohdanbirdie commented:
Any updates on this?

deftomat commented:
@bohdanbirdie We still waiting for the final review.  🤔

deftomat commented:
@ianschmitz Hi, any news on this?

ianschmitz commented:
Hey @deftomat, sorry it's been a busy few weeks for me.

@Timer and I were thinking we should probably add a test to ensure the build overlay is displayed after the async check.

deftomat commented:
@ianschmitz ok, I will try to create something.

deftomat commented:
@ianschmitz I created a test which will check overlay and console output for TypeScript error.

TravisCI is failing due to `Navigation Timeout Exceeded: 30000ms exceeded` in `fixtures/mjs-support/index.test.js`.

deftomat commented:
Failing TravisCI fixed by increased Puppeteer navigation timeouts.

ianschmitz commented:
> I created a test which will check overlay and console output for TypeScript error.

Thanks looks really good!

> Failing TravisCI fixed by increased Puppeteer navigation timeouts.

I'm not seeing this failure on master which is a little concerning. Looks like the default timeout is [30 seconds](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagegotourl-options). I only see the one build failing on this PR but i may have missed if there were more.

Maybe you would like to try reverting the timeout changes to use the default timeout and see if it passes? If that doesn't work then perhaps `await testSetup.scripts.start()` is resolving too soon for some reason?

deftomat commented:
@ianschmitz ok, I reverted it and it is ok. However, now I have error in overlay test in one build. It's quite strange as other builds are ok 😞

ianschmitz commented:
Yeah that's strange. It looks like the query only picked up part of the error message?

deftomat commented:
Yes

deftomat commented:
Everything is OK now! 🎉

Timer commented on packages/react-dev-utils/webpackHotDevClient.js:
```diff
@@ -140,19 +140,15 @@ function handleWarnings(warnings) {
   // Attempt to apply hot updates or reload.
   if (isHotUpdate) {
     tryApplyUpdates(function onSuccessfulHotUpdate() {
-      // Only print warnings if we aren't refreshing the page.
```
This is the only thing I'm on the fence about, what's the reason for this change? Can you please add the reasoning in the code?

deftomat replied:
The problem is that we need to print warnings even when there is not update in code. 
Without this change, https://github.com/facebook/create-react-app/blob/68e2f50dacec6205c42a108b6eecfc7f6d99e459/packages/react-dev-utils/webpackHotDevClient.js#L235-L237 will cancel it as `__webpack_hash__` didn't change.

johnnyreilly commented:
@deftomat - great work! I don't know if you saw https://github.com/facebook/create-react-app/issues/6194 - there's a new way to speed up incremental build speed in town!

@ianschmitz said:

> We would absolutely consider upgrading. This would go nicely with #5903

Would you consider folding this upgrade into your PR?

johnnyreilly commented:
Thought you'd be pleased 😉

@Timer can you help out please? 

We'd discussed previously making any changes required to `fork-ts-checker-webpack-plugin` directly. I'm not sure if any remain to be applied but I'm more than happy to help make that happen if further changes are required.

Or maybe we're just safe to switch? Seem to remember a discussion which concluded with that...

johnnyreilly commented:
> @johnnyreilly Migration was quite easy, but I'm not sure why it was necessary for me to remove these two lines

That's surprising to me as the projects I'm working on have both those set in the `tsconfig.json`. We don't pass the compiler options explicitly. Do you get errors if you use a `tsconfig.json` file?

`useTypescriptIncrementalApi: true` dropped incremental compilation from 12 seconds to 2 on a big project that I work on. The speed difference is *immense*.

I always run with `async: true` as well but I don't do anything as fancy as this PR does - I just watch the console and use this toast notifier: https://github.com/johnnyreilly/fork-ts-checker-notifier-webpack-plugin

Timer commented:
> Congratulations on the engagement @Timer!

Thank you 🎉 !

---

@johnnyreilly it looks like the `peerDependencies` are still present which was our main hesitation.

johnnyreilly commented:
Hey @Timer,

The latest release of plugin has just shipped - with `peerDependencies` dropped:

https://github.com/Realytics/fork-ts-checker-webpack-plugin/releases/tag/v1.0.0-alpha.3

To remove any thoughts of "wait, just hold on a minute.... he said `@next` - I am not that bleeding edge." Well, it's not like that. Don't be scared. 😄 

`fork-ts-checker-webpack-plugin` has merely been updated for webpack 5 (which is in alpha) and the `@next` reflects that. To be clear, the `@next` version of the plugin still supports (remarkably!) webpack 2, 3 and 4 as well as 5 alpha. Users of current and historic versions of webpack should feel safe using the `@next` version; for webpack 2, 3 and 4 expect stability. webpack 5 users should expect potential changes to align with webpack 5 as it progresses.

deftomat commented:
Alright, so I updated it to the latest `fork-ts-checker-webpack-plugin` and updated a way, how we emit the errors/warnings.

Now, the TypeScript errors are send into browser together with updated code when TypeScript check is faster than Babel. Otherwise, we send an updated code, show *"Files successfully emitted, waiting for typecheck results..."* in terminal and then send TypeScript errors as soon we received one.

However, there are a few issues:
1. It was necessary to remove [```moduleResolution: 'node'```](https://github.com/facebook/create-react-app/pull/5903/files#diff-dc0c4e7c623b73660da1809fc60cf6baL627) from `fork-ts-checker` config. I'm not sure why as my `tsconfig.json` already contains this option. Maybe it is a bug, any idea @johnnyreilly ?
  This brings me to the question: Why there are these compilerOptions anyway? We already have a check which will alter an existing `tsconfig.json` to match these options. I don't see any reason, so I removed all options for now. 
2. I'm not sure how tests are working as changing `fork-ts-checker-webpack-plugin-alt` to `fork-ts-checker-webpack-plugin` broke the most of them as they failed to find a typescript module. Maybe it is connected to [this option](https://github.com/facebook/create-react-app/blob/048822802ffd5d3db84b660f2f86d22117bc7e0e/packages/react-scripts/config/webpack.config.js#L619), which is probably supported just by the alternative (alt) implementation. When I add typescript into dependencies, then all tests pass.

REMINDER: We probably want to run `compile:lockfile` before merging this PR.

johnnyreilly commented:
> 1. It was necessary to remove `moduleResolution: 'node'` from `fork-ts-checker config`. I'm not sure why as my `tsconfig.json` already contains this option. Maybe it is a bug, any idea @johnnyreilly ?

I don't think it's a bug. It could be worth you looking at this comment by @0xorial here: https://github.com/namics/webpack-config-plugins/pull/43#issuecomment-455114496 - namely that `tsconfig.json` is loaded a different way with the incremental API.

> This brings me to the question: Why there are these compilerOptions anyway? We already have a check which will alter an existing tsconfig.json to match these options. I don't see any reason, so I removed all options for now.

I agree with this - I'm not sure what it was for either.

> I'm not sure how tests are working as changing fork-ts-checker-webpack-plugin-alt to fork-ts-checker-webpack-plugin broke the most of them as they failed to find a typescript module. Maybe it is connected to this option, which is probably supported just by the alternative (alt) implementation. When I add typescript into dependencies, then all tests pass.

I'm guessing this is because CRA doesn't want to force a depency on TypeScript given not everyone uses it. Either way, I'd be happy for that option to be supported by `fork-ts-checker-webpack-plugin`. If you'd like to submit a PR I can get it merged and shipped in pretty short order.

johnnyreilly commented:
PR to add `typescript` path to options here: https://github.com/Realytics/fork-ts-checker-webpack-plugin/pull/204/

deftomat commented:
@johnnyreilly Man, looks like we don't need to do anything. You will do it for us 😎 

Seriously, thanks for a really quick and detailed responses and PR. I will give it a try.

johnnyreilly commented:
Sorry about that - I'll take a look.

johnnyreilly commented:
It's published now: https://www.npmjs.com/package/fork-ts-checker-webpack-plugin/v/1.0.0-alpha.6

mikew commented:
I've been playing with branch combined with tsc and it's real fast. I did have to do some things to get it all working, I'm not sure if it's due to how I've added my branch of CRA to a project or using tsc instead of babel. Not sure if you're interested, but I had to do the following:

- Pass `fork-ts-checker-webpack-plugin` from `react-scripts` to `WebpackDevServerUtils.createCompiler`. Without this `forkTsCheckerWebpackPlugin.getCompilerHooks(compiler).receive` was never being called.
- Add the concept of "waiting for async errors / warnings" to `webpackHotDevClient`. This required `WebpackDevServerUtils` sending extra signals to `webpackHotDevClient`,
- Rework `compiler.hooks.compilation` / `compiler.hooks.done`. The `await` in `compiler.hooks.compilation` seemed to make the modifications to `compilation.errors` / `compilation.warnings` not appear in `compiler.hooks.done`, so `statsData` wasn't really necessary.

deftomat commented:
@mikew I'm pretty sure that we still want to use Babel.
> Rework compiler.hooks.compilation / compiler.hooks.done. The await in compiler.hooks.compilation seemed to make the modifications to compilation.errors / compilation.warnings not appear in compiler.hooks.done, so statsData wasn't really necessary.

I don't understand. 😕

deftomat commented:
Alright, I've updated fork-ts-checker to the latest version and all tests passed.
To make sure that everything is working as expected, I've merged my branch with `master` and it stops reporting problems.

Looks like errors pushed into compilation in `compiler.hooks.compilation` are ignored as observed by @mikew. 😞

deftomat commented:
No, no `tsc` experiments. 

Basically this branch merged with master and `fork-ts-checker` updated to alpha.6. Nothing more.

johnnyreilly commented:
Cool. Are you going to push your commit which update to alpha.6? It would be cool to see the PR tests go green 😊

deftomat commented:
Yes, I just need to resolve the issue which was introduced by merging this branch with `master`. Errors pushed asynchronly into compilation are ignored by webpack. This was not a case before merging it with `master` 😞

johnnyreilly commented:
Cool given it used to work I guess my `tapPromise` suggestion shouldn't be required...

deftomat commented:
Alright, most of the builds are green.

But `TEST_SUTE=install` is failing on the latest Node because it cannot find `react-dev-utils/crossSpawn`.

Any idea why? The module is in place (I've merged my branch with `master`)

Also, `TEST_SUITE=behavior` is failing too on the latest Node. Probably because of the same issue.

mikew commented:
Sorry, I didn't mean to derail the thread like that. Just wanted to point out some issues I was having that could potentially pop up (and it looks like it has). No intention of adding TSC to this, was just experimenting on my end.

mikew commented:
Nothing can be done with `compiler.hooks.done`, since it's a `SyncHook`. It doesn't have `tapPromise` or `tapAsync`.

Using `module.hot.accept` definitely helps with the issue of type errors not showing up. My branch still covers that. Here's changes applied to this branch: https://github.com/mikew/create-react-app/commit/f0e1b1bb30dbeeeb30f4fdf6cb0823a430d9b71c

deftomat commented:
@mikew I’m not sure what you are trying to do. Async errors are correctly rendered as tests passed.
Also, we want to apply hot updates even when we are waiting for typecheck. That’s the whole purpose of this PR.

The only issue which we currently have is that tests are not passing on the latest Node. And it’s probably not related to this.

Also, according to docs,  ‘compiler.hooks.done’ IS an async hook.

deftomat commented:
I looked at other PRs and they are failing with the same error message: `Cannot find module 'react-dev-utils/crossSpawn'`

So, I think that we are ready for the final review after someone from maintainers resolve this error. 🙂

mikew commented:
> Async errors are correctly rendered as tests passed.

This is what I'm seeing using this PR:

[![https://thumbs.gfycat.com/InsecureFirstFly-small.gif](https://thumbs.gfycat.com/InsecureFirstFly-small.gif)](https://gfycat.com/InsecureFirstFly)

Note the errors in the console but none in the browser (you may see them flash for a second before disappearing). Here it is with my changes:

[![https://thumbs.gfycat.com/EnchantedIckyIbex-small.gif](https://thumbs.gfycat.com/EnchantedIckyIbex-small.gif)](https://gfycat.com/EnchantedIckyIbex)

> Also, we want to apply hot updates even when we are waiting for typecheck.

For sure, and I fixed that shortly after in another commit.

Timer commented:
@mikew thanks for testing this. The behavior we want is definitely what you're displaying in the second picture (but I thought the PR worked liked that).

deftomat commented:
@mikew Just to be sure, are you testing it with `module.hot` export ? Or `react-hot-loaded` ?

Timer commented:
We need to make sure this works with `module.hot`.

deftomat commented:
Yes, definitely. I just discovered that for me, it works for 80% of times. However, sometimes, error just won't show up. Not sure why, maybe because typecheck is now so fast in small projects. It works flawlessly when TypeScript was always slower then Babel....

@mikew I really appreciate that you are trying to fix this but could you explain to us what needs to be changed and why. Just a information that your branch has a correct behavior is not really helpful.

deftomat commented:
@mikew What about this https://github.com/facebook/create-react-app/pull/5903/commits/321d960a67ad3dba8633adfe3292f14c281db7d8 ?

Turns out that there is already the `hasCompileErrors` flag in place. The only change is that now we are checking if it is `false` when we are going to dismiss error overlay.

johnnyreilly commented:
> (I'm "blaming" you @johnnyreilly 😉 You guys just make it too fast 👍)

I'll take that! 😊

deftomat commented:
@Timer Well, we will never be sure that it is always that fast. As mentioned by @johnnyreilly, for him, it could still took 2 seconds in large project. I think that when we have an opportunity to be sure, that TypeScript will not slow down anyone, then I think we should do it.

Also, after 2 months, it turns out that the added complexity is just 2 IFs in `WebpackDevServerUtils.js`. Other changes are just improvements to existing features, like better TypeScript message formatter, and making sure that errors are not accidentally dismissed.

Timer commented:
So does it work "perfectly now", or is there still an edge case? If this is ready, I can give it a test.

deftomat commented:
I think that it finally works as expected.

mikew commented:
> If it's really that fast now, should we add this async complexity?

As @deftomat mentioned, it's not a lot of complexity. And it opens the door for these async errors / warnings for other plugins, which might be a win down the line. But yeah, the real gains are from using the incremental API, which works without `async`.

> So does it work "perfectly now", or is there still an edge case?

It's working as expected for me.

johnnyreilly commented:
I'd ❤️ this to be merged....

ianschmitz commented:
I'll give it another spin tomorrow night, and barring any glaring issues we will get it in. Thanks for your patience!

ianschmitz commented:
Thanks everyone for your help! I gave it a spin with a couple of my projects and it seems to be working really nice. TypeScript is now super fast! 

I'll see if we can cut a release in the next day or two. 🎉

ianschmitz commented:
[v2.1.4](https://github.com/facebook/create-react-app/releases/tag/v2.1.4) is now out!

johnnyreilly commented:
Heya, 

I can see that https://github.com/facebook/create-react-app/pull/6395 has been merged but it also says here that #5903 was reverted. 

Could you clarify where are things now please? I'm keen to get the improved TypeScript experience out there in a way that doesn't impede people who don't want to use TypeScript...

ianschmitz commented:
Hey @johnnyreilly. Yes that's correct we've temporarily reverted this PR. I mistakenly published `react-dev-utils` as a patch level increase when it included the breaking change to https://github.com/facebook/create-react-app/pull/5903/files#diff-6736a5f1ee20761aaa6d0f4a09249e30, as well as a missing dependency of `fork-ts-checker-webpack-plugin` in the `package.json` of this package. This caused issues for other packages that rely on the public API of `react-dev-utils`.

The best short term solution was to publish a newer patch version to revert these changes so downstream consumers of `react-dev-utils` aren't in a broken state. We are very keen to get some form of this back into master, but just want to be careful of the breaking changes in the next revision.

johnnyreilly commented:
Great - thanks for clarifying!

timothyallan commented:
My day today: Typescript compiling slooooow.....  Typescript compiling fast!!!....  Typescript compiling slooooow.

jakobthomasson commented:
So when is this PR gonna be merged into master again :(

deftomat commented:
@jakobthomasson See https://github.com/facebook/create-react-app/pull/6406 for any progress
